{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# cupy test\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'nvcc' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mnvcc\u001b[49m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m-\u001b[39mversion\n",
      "\u001b[0;31mNameError\u001b[0m: name 'nvcc' is not defined"
     ]
    }
   ],
   "source": [
    "nvcc --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mDEPRECATION: Loading egg at /opt/miniconda/envs/r42/lib/python3.12/site-packages/example-0.1-py3.12-linux-x86_64.egg is deprecated. pip 25.1 will enforce this behaviour change. A possible replacement is to use pip for package installation. Discussion can be found at https://github.com/pypa/pip/issues/12330\u001b[0m\u001b[33m\n",
      "\u001b[0mCollecting cupy\n",
      "  Downloading cupy-13.3.0.tar.gz (3.4 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.4/3.4 MB\u001b[0m \u001b[31m1.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: numpy<2.3,>=1.22 in /opt/miniconda/envs/r42/lib/python3.12/site-packages (from cupy) (1.26.4)\n",
      "Collecting fastrlock>=0.5 (from cupy)\n",
      "  Using cached fastrlock-0.8.3-cp312-cp312-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_28_x86_64.whl.metadata (7.7 kB)\n",
      "Using cached fastrlock-0.8.3-cp312-cp312-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_28_x86_64.whl (53 kB)\n",
      "Building wheels for collected packages: cupy\n",
      "  Building wheel for cupy (setup.py) ... \u001b[?25lerror\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mpython setup.py bdist_wheel\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[65 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m Generating cache key from header files...\n",
      "  \u001b[31m   \u001b[0m Cache key (1610 files matching /tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/cupy/_core/include/**): 3cb22ea53d3f485714e9bca85380aa4e10974776\n",
      "  \u001b[31m   \u001b[0m Clearing directory: /tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/cupy/.data\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m -------- Configuring Module: cuda --------\n",
      "  \u001b[31m   \u001b[0m /tmp/tmpy1qixkfj/a.cpp:1:10: fatal error: cublas_v2.h: No such file or directory\n",
      "  \u001b[31m   \u001b[0m     1 | #include <cublas_v2.h>\n",
      "  \u001b[31m   \u001b[0m       |          ^~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m compilation terminated.\n",
      "  \u001b[31m   \u001b[0m command '/opt/miniconda/bin/x86_64-conda-linux-gnu-c++' failed with exit code 1\n",
      "  \u001b[31m   \u001b[0m /tmp/tmpb_niekdh/a.cpp:2:18: fatal error: cuda_runtime_api.h: No such file or directory\n",
      "  \u001b[31m   \u001b[0m     2 |         #include <cuda_runtime_api.h>\n",
      "  \u001b[31m   \u001b[0m       |                  ^~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m compilation terminated.\n",
      "  \u001b[31m   \u001b[0m **************************************************\n",
      "  \u001b[31m   \u001b[0m *** WARNING: Cannot check compute capability\n",
      "  \u001b[31m   \u001b[0m command '/opt/miniconda/bin/x86_64-conda-linux-gnu-c++' failed with exit code 1\n",
      "  \u001b[31m   \u001b[0m **************************************************\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m ************************************************************\n",
      "  \u001b[31m   \u001b[0m * CuPy Configuration Summary                               *\n",
      "  \u001b[31m   \u001b[0m ************************************************************\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Build Environment:\n",
      "  \u001b[31m   \u001b[0m   Include directories: ['/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/cupy/_core/include/cupy/_cccl/libcudacxx', '/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/cupy/_core/include/cupy/_cccl/thrust', '/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/cupy/_core/include/cupy/_cccl/cub', '/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/cupy/_core/include']\n",
      "  \u001b[31m   \u001b[0m   Library directories: []\n",
      "  \u001b[31m   \u001b[0m   nvcc command       : (not found)\n",
      "  \u001b[31m   \u001b[0m   hipcc command      : (not found)\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Environment Variables:\n",
      "  \u001b[31m   \u001b[0m   CFLAGS          : -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/miniconda/include\n",
      "  \u001b[31m   \u001b[0m   LDFLAGS         : -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/opt/miniconda/lib -Wl,-rpath-link,/opt/miniconda/lib -L/opt/miniconda/lib\n",
      "  \u001b[31m   \u001b[0m   LIBRARY_PATH    : :/opt/miniconda/envs/r42/lib/R/library/RcppEigen/lib:/opt/miniconda/envs/r42/lib/R/library/Rcpp/lib:/opt/miniconda/envs/r42/lib/R/include\n",
      "  \u001b[31m   \u001b[0m   CUDA_PATH       : (none)\n",
      "  \u001b[31m   \u001b[0m   NVCC            : (none)\n",
      "  \u001b[31m   \u001b[0m   HIPCC           : (none)\n",
      "  \u001b[31m   \u001b[0m   ROCM_HOME       : (none)\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Modules:\n",
      "  \u001b[31m   \u001b[0m   cuda      : No\n",
      "  \u001b[31m   \u001b[0m     -> Include files not found: ['cublas_v2.h', 'cuda.h', 'cuda_profiler_api.h', 'cuda_runtime.h', 'cufft.h', 'curand.h', 'cusparse.h']\n",
      "  \u001b[31m   \u001b[0m     -> Check your CFLAGS environment variable.\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m ERROR: CUDA could not be found on your system.\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m HINT: You are trying to build CuPy from source, which is NOT recommended for general use.\n",
      "  \u001b[31m   \u001b[0m       Please consider using binary packages instead.\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Please refer to the Installation Guide for details:\n",
      "  \u001b[31m   \u001b[0m https://docs.cupy.dev/en/stable/install.html\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m ************************************************************\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"<string>\", line 2, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"<pip-setuptools-caller>\", line 34, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/setup.py\", line 95, in <module>\n",
      "  \u001b[31m   \u001b[0m     ext_modules = cupy_setup_build.get_ext_modules(True, ctx)\n",
      "  \u001b[31m   \u001b[0m                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  \u001b[31m   \u001b[0m   File \"/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/install/cupy_builder/cupy_setup_build.py\", line 514, in get_ext_modules\n",
      "  \u001b[31m   \u001b[0m     extensions = make_extensions(ctx, compiler, use_cython)\n",
      "  \u001b[31m   \u001b[0m                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  \u001b[31m   \u001b[0m   File \"/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/install/cupy_builder/cupy_setup_build.py\", line 363, in make_extensions\n",
      "  \u001b[31m   \u001b[0m     raise Exception('Your CUDA environment is invalid. '\n",
      "  \u001b[31m   \u001b[0m Exception: Your CUDA environment is invalid. Please check above error log.\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[31m  ERROR: Failed building wheel for cupy\u001b[0m\u001b[31m\n",
      "\u001b[0m\u001b[?25h  Running setup.py clean for cupy\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mpython setup.py clean\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[65 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m Generating cache key from header files...\n",
      "  \u001b[31m   \u001b[0m Cache key (1610 files matching /tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/cupy/_core/include/**): 3cb22ea53d3f485714e9bca85380aa4e10974776\n",
      "  \u001b[31m   \u001b[0m Clearing directory: /tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/cupy/.data\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m -------- Configuring Module: cuda --------\n",
      "  \u001b[31m   \u001b[0m /tmp/tmpzr0p9jo7/a.cpp:1:10: fatal error: cublas_v2.h: No such file or directory\n",
      "  \u001b[31m   \u001b[0m     1 | #include <cublas_v2.h>\n",
      "  \u001b[31m   \u001b[0m       |          ^~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m compilation terminated.\n",
      "  \u001b[31m   \u001b[0m command '/opt/miniconda/bin/x86_64-conda-linux-gnu-c++' failed with exit code 1\n",
      "  \u001b[31m   \u001b[0m /tmp/tmp5icwqfz4/a.cpp:2:18: fatal error: cuda_runtime_api.h: No such file or directory\n",
      "  \u001b[31m   \u001b[0m     2 |         #include <cuda_runtime_api.h>\n",
      "  \u001b[31m   \u001b[0m       |                  ^~~~~~~~~~~~~~~~~~~~\n",
      "  \u001b[31m   \u001b[0m compilation terminated.\n",
      "  \u001b[31m   \u001b[0m **************************************************\n",
      "  \u001b[31m   \u001b[0m *** WARNING: Cannot check compute capability\n",
      "  \u001b[31m   \u001b[0m command '/opt/miniconda/bin/x86_64-conda-linux-gnu-c++' failed with exit code 1\n",
      "  \u001b[31m   \u001b[0m **************************************************\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m ************************************************************\n",
      "  \u001b[31m   \u001b[0m * CuPy Configuration Summary                               *\n",
      "  \u001b[31m   \u001b[0m ************************************************************\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Build Environment:\n",
      "  \u001b[31m   \u001b[0m   Include directories: ['/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/cupy/_core/include/cupy/_cccl/libcudacxx', '/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/cupy/_core/include/cupy/_cccl/thrust', '/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/cupy/_core/include/cupy/_cccl/cub', '/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/cupy/_core/include']\n",
      "  \u001b[31m   \u001b[0m   Library directories: []\n",
      "  \u001b[31m   \u001b[0m   nvcc command       : (not found)\n",
      "  \u001b[31m   \u001b[0m   hipcc command      : (not found)\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Environment Variables:\n",
      "  \u001b[31m   \u001b[0m   CFLAGS          : -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/miniconda/include\n",
      "  \u001b[31m   \u001b[0m   LDFLAGS         : -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/opt/miniconda/lib -Wl,-rpath-link,/opt/miniconda/lib -L/opt/miniconda/lib\n",
      "  \u001b[31m   \u001b[0m   LIBRARY_PATH    : :/opt/miniconda/envs/r42/lib/R/library/RcppEigen/lib:/opt/miniconda/envs/r42/lib/R/library/Rcpp/lib:/opt/miniconda/envs/r42/lib/R/include\n",
      "  \u001b[31m   \u001b[0m   CUDA_PATH       : (none)\n",
      "  \u001b[31m   \u001b[0m   NVCC            : (none)\n",
      "  \u001b[31m   \u001b[0m   HIPCC           : (none)\n",
      "  \u001b[31m   \u001b[0m   ROCM_HOME       : (none)\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Modules:\n",
      "  \u001b[31m   \u001b[0m   cuda      : No\n",
      "  \u001b[31m   \u001b[0m     -> Include files not found: ['cublas_v2.h', 'cuda.h', 'cuda_profiler_api.h', 'cuda_runtime.h', 'cufft.h', 'curand.h', 'cusparse.h']\n",
      "  \u001b[31m   \u001b[0m     -> Check your CFLAGS environment variable.\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m ERROR: CUDA could not be found on your system.\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m HINT: You are trying to build CuPy from source, which is NOT recommended for general use.\n",
      "  \u001b[31m   \u001b[0m       Please consider using binary packages instead.\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Please refer to the Installation Guide for details:\n",
      "  \u001b[31m   \u001b[0m https://docs.cupy.dev/en/stable/install.html\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m ************************************************************\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"<string>\", line 2, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"<pip-setuptools-caller>\", line 34, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/setup.py\", line 95, in <module>\n",
      "  \u001b[31m   \u001b[0m     ext_modules = cupy_setup_build.get_ext_modules(True, ctx)\n",
      "  \u001b[31m   \u001b[0m                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  \u001b[31m   \u001b[0m   File \"/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/install/cupy_builder/cupy_setup_build.py\", line 514, in get_ext_modules\n",
      "  \u001b[31m   \u001b[0m     extensions = make_extensions(ctx, compiler, use_cython)\n",
      "  \u001b[31m   \u001b[0m                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  \u001b[31m   \u001b[0m   File \"/tmp/pip-install-89qzriha/cupy_88e01f41d1e94690a57fed24234b2af4/install/cupy_builder/cupy_setup_build.py\", line 363, in make_extensions\n",
      "  \u001b[31m   \u001b[0m     raise Exception('Your CUDA environment is invalid. '\n",
      "  \u001b[31m   \u001b[0m Exception: Your CUDA environment is invalid. Please check above error log.\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[31m  ERROR: Failed cleaning build dir for cupy\u001b[0m\u001b[31m\n",
      "\u001b[0mFailed to build cupy\n",
      "\u001b[31mERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (cupy)\u001b[0m\u001b[31m\n",
      "\u001b[0mNote: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "pip install cupy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "import cupy as cp\n",
    "\n",
    "# 创建两个随机矩阵\n",
    "matrix_a = cp.random.rand(3, 4)  # 3x4 矩阵\n",
    "matrix_b = cp.random.rand(4, 5)  # 4x5 矩阵\n",
    "\n",
    "# 矩阵乘法\n",
    "result = cp.dot(matrix_a, matrix_b)\n",
    "\n",
    "# 打印结果\n",
    "print(\"Matrix A:\")\n",
    "print(matrix_a)\n",
    "\n",
    "print(\"\\nMatrix B:\")\n",
    "print(matrix_b)\n",
    "\n",
    "print(\"\\nResult of Matrix Multiplication:\")\n",
    "print(result)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r42",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
